### PR TITLE
rpc/request: delay parameter unmarshaling

### PR DESCRIPTION
**Pull request by fyrchik**
It is rather costly to try to unmarshal many structs in order.

Signed-off-by: Evgeniy Stratonikov <evgeniy@nspcc.ru>

### CODE REVIEW

File: pkg/rpc/request/param.go [hunk 1]
```diff
@@ -264,83 +310,47 @@ func (p Param) GetSignersWithWitnesses() ([]transaction.Signer, []transaction.Wi
...
+		p.Value = t
+	case float64: // unexpected because of `UseNumber`.
+		panic("unexpected")
+	case json.Number:
+		value, err := strconv.ParseInt(string(t), 10, 64)
+		if err != nil {
+			return err
+		}
+		p.Type = NumberT
+		p.Value = int(value)
```
Comment (roman-khimov) [inline pkg/rpc/request/param.go:346]: CodeQL is angry about this.

File: pkg/rpc/request/param.go [hunk 2]
```diff
@@ -264,83 +310,47 @@ func (p Param) GetSignersWithWitnesses() ([]transaction.Signer, []transaction.Wi
...
-		{SignerWithWitnessT, &signerWithWitnessAux{}},
-		{ArrayT, &[]Param{}},
-	}
-
-	if bytes.Equal(data, []byte("null")) {
-		p.Type = defaultT
-		return nil
+	r := bytes.NewReader(data)
+	jd := json.NewDecoder(r)
+	jd.DisallowUnknownFields()
```
Comment (roman-khimov) [inline pkg/rpc/request/param.go:315]: Maybe not really needed, we no longer unmarshal structs here.

File: pkg/rpc/server/server.go [hunk 3]
```diff
@@ -1496,28 +1497,41 @@ func (s *Server) subscribe(reqParams request.Params, sub *subscriber) (interface
...
 		case response.ExecutionEventID:
-			if p.Type != request.ExecutionFilterT {
-				return nil, response.ErrInvalidParams
-			}
-		case response.NotaryRequestEventID:
-			if p.Type != request.TxFilterT {
-				return nil, response.ErrInvalidParams
+			flt := new(request.ExecutionFilter)
+			err = jd.Decode(flt)
+			if err == nil && (flt.State == "HALT" || flt.State == "FAULT") {
```
Comment (roman-khimov) [inline pkg/rpc/server/server.go:1525]: Do we have tests for this?

**Review by roman-khimov — CHANGES_REQUESTED** [review]

File: pkg/rpc/request/param.go [hunk 4]
```diff
@@ -219,11 +238,38 @@ func (p *Param) GetBytesBase64() ([]byte, error) {
 }
 
 // GetSignerWithWitness returns SignerWithWitness value of the parameter.
-func (p Param) GetSignerWithWitness() (SignerWithWitness, error) {
+func (p *Param) GetSignerWithWitness() (SignerWithWitness, error) {
 	c, ok := p.Value.(SignerWithWitness)
+	if ok {
```
Comment (AnnaShaleva) [inline pkg/rpc/request/param.go:243]: Isn't then `p.Value` always a raw message?

File: pkg/rpc/request/param.go [hunk 5]
```diff
@@ -134,9 +134,19 @@ func (p *Param) GetArray() ([]Param, error) {
 		return nil, errMissingParameter
 	}
 	a, ok := p.Value.([]Param)
+	if ok {
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/rpc/request/param.go:137]: Fixed.

File: pkg/rpc/request/param.go [hunk 6]
```diff
@@ -264,83 +311,39 @@ func (p Param) GetSignersWithWitnesses() ([]transaction.Signer, []transaction.Wi
...
+			p.Type = ArrayT
+			p.Value = arr
+		} else {
+			p.Type = defaultT
+			p.Value = json.RawMessage(data)
 		}
+	case bool:
+		p.Type = BooleanT
+		p.Value = t
+	case float64: // unexpected because of `UseNumber`.
```
Comment (roman-khimov) [inline pkg/rpc/request/param.go:337]: Now you don't `UseNumber` and the code is even more dangerous.

File: pkg/rpc/request/param.go (hunk 4 above)
Comment (fyrchik) [reply pkg/rpc/request/param.go:244]: It is set right before the return.

**Review by roman-khimov — COMMENTED** [review]
Squash?

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**